Source/FieldSolver/ElectrostaticSolvers/EffectivePotentialES.cpp

```diff
@@ -0,0 +1,257 @@
+/* Copyright 2024 The WarpX Community
+ *
+ * This file is part of WarpX.
+ *
+ * Authors: Roelof Groenewald (TAE Technologies)
+ *
+ * License: BSD-3-Clause-LBNL
+ */
+
+#include "EffectivePotentialES.H"
+#include "Fluids/MultiFluidContainer_fwd.H"
+#include "EmbeddedBoundary/Enabled.H"
+#include "Fields.H"
+#include "Particles/MultiParticleContainer_fwd.H"
+#include "Utils/Parser/ParserUtils.H"
+#include "WarpX.H"
+
+using namespace amrex;
+
+void EffectivePotentialES::InitData() {
+    auto & warpx = WarpX::GetInstance();
+    m_poisson_boundary_handler->DefinePhiBCs(warpx.Geom(0));
+}
+
+void EffectivePotentialES::ComputeSpaceChargeField (
+    ablastr::fields::MultiFabRegister& fields,
+    MultiParticleContainer& mpc,
+    [[maybe_unused]] MultiFluidContainer* mfl,
+    int max_level)
+{
+    WARPX_PROFILE("EffectivePotentialES::ComputeSpaceChargeField");
+
+    using ablastr::fields::MultiLevelScalarField;
+    using ablastr::fields::MultiLevelVectorField;
+    using warpx::fields::FieldType;
+
+    // grab the simulation fields
+    const MultiLevelScalarField rho_fp = fields.get_mr_levels(FieldType::rho_fp, max_level);
+    const MultiLevelScalarField rho_cp = fields.get_mr_levels(FieldType::rho_cp, max_level);
+    const MultiLevelScalarField phi_fp = fields.get_mr_levels(FieldType::phi_fp, max_level);
+    const MultiLevelVectorField Efield_fp = fields.get_mr_levels_alldirs(FieldType::Efield_fp, max_level);
+
+    mpc.DepositCharge(rho_fp, 0.0_rt);
+    if (mfl) {
+        const int lev = 0;
+        mfl->DepositCharge(fields, *rho_fp[lev], lev);
+    }
+
+    // Apply filter, perform MPI exchange, interpolate across levels
+    const Vector<std::unique_ptr<MultiFab> > rho_buf(num_levels);
+    auto & warpx = WarpX::GetInstance();
+    warpx.SyncRho( rho_fp, rho_cp, amrex::GetVecOfPtrs(rho_buf) );
+
+#ifndef WARPX_DIM_RZ
+    for (int lev = 0; lev < num_levels; lev++) {
+        // Reflect density over PEC boundaries, if needed.
+        warpx.ApplyRhofieldBoundary(lev, rho_fp[lev], PatchType::fine);
+    }
+#endif
+
+    // set the boundary potentials appropriately
+    setPhiBC(phi_fp, warpx.gett_new(0));
+
+    // perform phi calculation
+    computePhi(rho_fp, phi_fp);
+
+    // Compute the electric field. Note that if an EB is used the electric
+    // field will be calculated in the computePhi call.
+    const std::array<Real, 3> beta = {0._rt};
+    if (!EB::enabled()) { computeE( Efield_fp, phi_fp, beta ); }
+}
+
+void EffectivePotentialES::computePhi (
+    ablastr::fields::MultiLevelScalarField const& rho,
+    ablastr::fields::MultiLevelScalarField const& phi) const
+{
+    // Calculate the mass enhancement factor:
+    // The "sigma" multifab stores the dressing of the Poisson equation. It
+    // is a cell-centered multifab.
```

Comment:
Could you again add a reference to Dan Barnes paper here, just to remind whoever reads the code where this comes from?